### PR TITLE
Make DPI function signatures standard compliant

### PR DESCRIPTION
**Pull request by gezalore**
This PR adresses #2207, but also makes argument types, not just return types stantard compliant (it was simpler to do the args at the same time). There should also now be precise errors when a return type is invalid. The tests are a bit tedious, but I was unfamiliar with a lot of the internals, so I went a bit paranoid. The compiler internals should have gottensimpler though. Tests that I have deleted are coverted by the new ones. I was also a bit confused as of what copyright header is preferred for the tests, I'm happy for you to change it in any way if necessary.

For the actual fucntion signature now emitted see the `t_dpi_*type__Dpi.out` golden files. All of those matches the big 3 now.

### CODE REVIEW

File: src/astgen [hunk 1]
```diff
@@ -137,11 +137,24 @@ sub read_refs {
...
-sub open_file {
+sub open_output_file {
     my $filename = shift;
-    my $fh = IO::File->new($filename,"w") or die "%Error: $! $filename,";
-    print $fh '// Generated by astgen // -*- mode: C++; c-file-style: "cc-mode" -*-'."\n";
-    return $fh;
+    if (!defined($filename)) {
+        return \*STDOUT;
+    } else {
+        chmod 0660, $filename if $^O eq "linux" && -e $filename;
```
Comment (wsnyder) [inline src/astgen:145]: Why is the chmod needed?  This might remove world permission which might be desired.

File: src/astgen [hunk 2]
```diff
@@ -137,11 +137,24 @@ sub read_refs {
 
 #----------------------------------------------------------------------
```
Comment (wsnyder) [inline src/astgen:139]: I suspect these astgen changes aren't intended to be part of this - doesn't seem related.

File: test_regress/t/t_dpi_accessors.cpp [hunk 3]
```diff
@@ -246,7 +246,7 @@ int main() {
     for (int i = 0; !Verilated::gotFinish() && (i < 4); i++) {
         dut->clk = 1 - dut->clk;
         a = 1 - (int)a_read();
-        a_write(a);
+        a_write((const svBitVecVal*)&a);
```
Comment (wsnyder) [inline test_regress/t/t_dpi_accessors.cpp:249]: Please use C++ style cast if possible.

File: src/V3Task.cpp [hunk 4]
```diff
@@ -927,15 +915,23 @@ class TaskVisitor : public AstNVisitor {
...
+            if (nodep->dpiImport() || nodep->dpiExport()) {
+                AstBasicDType *bdtypep = portp->dtypep()->basicp();
+                if (!bdtypep->isDpiPrimitive()) {
+                    if (bdtypep->isDpiBitVec() && portp->width() > 32) {
+                        portp->v3error("DPI function may not return a > 32 bits wide type "
+                                       "other than basic types. (Make it a output instead.)");
+                    }
+                    if (bdtypep->isDpiLogicVec()) {
+                        portp->v3error("DPI function may not return a 4-state type "
+                                       "other than a single 'logic' (IEEE 1800-2017 35.5.5).");
```
Comment (wsnyder) [inline src/V3Task.cpp:927]: Nit, remove trailing period.

File: src/V3Task.cpp [hunk 5]
```diff
@@ -927,15 +915,23 @@ class TaskVisitor : public AstNVisitor {
...
-                && portp->dtypep()->basicp()->keyword().isDpiUnreturnable()) {
-                portp->v3error("DPI function may not return type "
-                               <<portp->basicp()->prettyTypeName()
-                               <<" (IEEE 1800-2017 35.5.5)");
+            if (nodep->dpiImport() || nodep->dpiExport()) {
+                AstBasicDType *bdtypep = portp->dtypep()->basicp();
+                if (!bdtypep->isDpiPrimitive()) {
+                    if (bdtypep->isDpiBitVec() && portp->width() > 32) {
+                        portp->v3error("DPI function may not return a > 32 bits wide type "
+                                       "other than basic types. (Make it a output instead.)");
```
Comment (wsnyder) [inline src/V3Task.cpp:923]: "other than basic types.\n"
     + V3Error::warnMore() + "... Suggest make it a output instead?");

similar below

**Comment by gezalore** [comment]
> Can you please add whatever ifdefs are needed to have these pass on the big 3? This is important for these (unlike most other tests) as the resulting ifdefs help indicate where the standard isn't so standard. Maybe now this will be behind us thanks to your work, but still would like them portable. Note boilerplate #ifdef stuff needed at the top of the .cpp file to get the right headers.

No worries, I will do that.

> I'd like Verilog files to use verilog-mode standard indents generally, iI'll reindent either in your repo or pre-merge for you unless you want to do it.

I'm happy to do it, though a one-liner for verilog-mode batch mode would be helpful, I don't normally use it.

File: test_regress/t/t_dpi_accessors.cpp (hunk 3 above)
Comment (gezalore) [reply test_regress/t/t_dpi_accessors.cpp:249]: The rest of that file uses C style casts, so I just copied what was already there. I will have a skim through.

File: src/astgen (hunk 1 above)
Comment (gezalore) [reply src/astgen:145]: If the output file already exists (as checked in the condition), i.e.: you already run astgen on a previous build, then it was previously made read only by astgen when it output it, so the open for write below would fail without the chmod to add w pemissions. I was under the impression that these are just internal file (and the astgen processed versions of those) used while building Verilator so the permissions should only mater to the dev (who is the user), but not externally.

File: src/astgen (hunk 2 above)
Comment (gezalore) [reply src/astgen:139]: Not really related to the DPI signatures. Just a general developer quality of life thing.. I edited the generated files too often after my editor opend them up with "Go to definition", silly me.

File: src/astgen (hunk 1 above)
Comment (wsnyder) [reply src/astgen:145]: Well same question then about the chmod to make it read only.  Don't think we should be messing with permissions at all.  If you're worried about editing it, your editor probably has a way to make a directory or filename read-only by default, or if you were using emacs we could add the default-to-read-only inside the file itself.

**Comment by wsnyder** [comment]
emacs --batch file.v --eval "(setq-default indent-tabs-mode nil)" -f verilog-batch-indent -f verilog-batch-auto

Pull/merge from what I just pushed to master, then try that on e.g. t_EXAMPLE.v and it should have no changes.

**Comment by wsnyder** [comment]
Also try clang-format on your test C code - note it'll muck up any "pretty" columns and always does preprocessor stuff wrong, so don't take it's opinion blindly.

**Comment by gezalore** [comment]
I have finished porting one of the tests to the big 3 (and as a result I now appreciate Verilator even more than before!), would you be happy with these changes to the test driver for Modelsim:
https://github.com/gezalore/verilator/commit/ef9c5b255d2f391518799871c7801a1e4eb81608
They just add a Verilog and a C preprocessor define. Feels like we should just always have these, but otherwise I can always just stick them in `ms_flags2` in the tests themselves, your choice. 

I am curious, are the tests regularly run against the big 3 in some dark cupboard somewhere?

**Comment by wsnyder** [comment]
The way you added those flags for modelsim seems best.  We don't run other simulators regularly, certainly there will be many failures to be cleaned up &  some tests have probably never been clean on others.

**Comment by wsnyder** [comment]
It looks like your modelsim changes is in a separate branch but builds on this pull - can you merge it in here please, and merge in master?  Also please update this message

                        portp->v3error("DPI function may not return a > 32 bits wide type "
                                       "other than basic types. (Make it a output instead.)");

To be ~~
                        portp->v3error("DPI function may not return a > 32 bit wide type "
                                       "other than basic types.\n"
                                       + V3Error::warnMore() + "... Suggest make it a parameter that is an output instead?");

I realize that what you copied it from had the bad wording, but think as previously worded "output" might be is confusing.

**Comment by wsnyder** [comment]
There's a second error message a bit below that should have similar treatment.

**Comment by gezalore** [comment]
Sure, I will do all that and will address all your earlier comments as well once I finished porting all the tests. Right now I just wanted to ask for guidance regarding testing on the big3 (which you have now provided). I’m sorry I wasn’t clear. I will ping here when I’m happy with the patch. Thank you again for the help.

File: src/astgen (hunk 1 above)
Comment (gezalore) [reply src/astgen]: No problem. I have removed tha changes to astgen from this PR

File: src/astgen (hunk 2 above)
Comment (gezalore) [reply src/astgen]: I have removed tha changes to astgen from this PR

File: test_regress/t/t_dpi_accessors.cpp (hunk 3 above)
Comment (gezalore) [reply test_regress/t/t_dpi_accessors.cpp:249]: I fixed all casts in that test in a separate commit in this PR.

File: src/V3Task.cpp (hunk 4 above)
Comment (gezalore) [reply src/V3Task.cpp:927]: Done

File: src/V3Task.cpp (hunk 5 above)
Comment (gezalore) [reply src/V3Task.cpp:923]: Done both as
```
  + V3Error::warnMore() + "... Suggest make it an output argument instead?");
```

**Comment by gezalore** [comment]
@wsnyder I think it's now ready for the next review. Please see above for the specific items.

The 4 new tests that are not expected to fail now all pass with the big 3 as well.

I don't think thera are any significant differences between the tools. Apart form the usual "X does not support type Y at all", the differences are:
- Modelsim/Questa declares SV `longint` as `int64_t` rather than the standard specified `long long`, and on the version of GCC that is bundled with it `int64_t` is `long int` rather than `long long`, so I needed to typedef around that.
- NC/Incisiv declares pass-by reference input arguments with a non-`const` pointrers while everyone else uses `const` qualified pointers. The version of NC I tried is from 2015, so this might have been fixed in newer releases.

**Comment by gezalore** [comment]
For posterity's sake, here are the exact tool versions I used:

> % irun -version
> TOOL:   irun    15.22-s012
> % vlog -version
> QuestaSim-64 vlog 10.7d Compiler 2019.02 Feb 15 2019
> % vcs -ID | grep "Compiler version"
> Compiler version = VCS Q-2020.03

File: test_regress/t/t_dpi_arg_input_type.v [hunk 6]
```diff
@@ -0,0 +1,917 @@
...
+   wire  _unused = &{1'b0, clk};
+`endif
+
+   // Legal input argument types for DPI functions
+
+   //////////////////////////////////////////////////////////////////////////////
+   // Type definitions
+   //////////////////////////////////////////////////////////////////////////////
+
+   // Basic types as per IEEE 1800-2017 35.5.6
```
Comment (gezalore) [inline test_regress/t/t_dpi_arg_input_type.v:45]: Forgot to mention: I applied clang-format and the verilog-mode batch indent as you suggested, though some stuff in the Verilog came out a bit odd looking, like these typedefs below.

**Comment by wsnyder** [comment]
1. Using somewhat newer NC they still don't do const nor integer.
2. Got compiler error for illegal to left shift a -1, I changed it to 0xffffffff.
3. Cleaned some spaces and merged outside github.

Thanks for your extraordinary efforts on this.
